protocol_generator/protocol_generator.cpp

```diff
@@ -0,0 +1,234 @@
+			d = top;
+			d.avail_coins.erase(std::begin(d.avail_coins));
+			s.push(d);
+		}
+		std::cout << target << ":" << ways << std::endl;
+		protocol_map[target] = chosen_protocols;
+	}
+	std::cout << ways << std::endl;
+	std::cout << "Size " << protocol_map.size() << "\n";
+  	std::cout << "Count " << protocol_map.count(2.0) << "\n";
+
+	// 1. Losowanko
+	// 1.1 losuj sumę - równa dystrybucja po wartości sumy
+	// 1.2 losuj pozycję
+	// 1.3 losuj permutacje
+	// 1.4 losuj czasy
+
+	std::set<std::vector<float>> protocol_set;
+
+	std::random_device rd;
+	std::mt19937 rng;
+	rng.seed(rd());
+	std::uniform_int_distribution<int> dist_sum(3, 40); 
+	int i = 0;
+	std::ofstream f("protocol_list_1.csv");
+	std::ofstream t("protocol_times_1.csv");
+	int file_count = 1;
+	while (i < 50000) {
+		// Ile linii na jeden plik
+		if (i > file_count * 1570) {
+			file_count++;
+			f.close();
+			t.close();
+			f.open("protocol_list_" + std::to_string(file_count) + ".csv");
+			t.open("protocol_times_" + std::to_string(file_count) + ".csv");
+		}
+		// Losuj sumę
+		float sum = dist_sum(rng) * 0.25;
+
+		std::vector<float> temp_protocol;
+
+		std::cout << "Suma " << sum << "\n";
+		std::cout << "Rozmiar " << protocol_map[sum].size() << "\n";
+		// Losuj pozycję
+		std::uniform_int_distribution<int> dist_pos(
+			0, protocol_map[sum].size() - 1);
+		int temp_pos = dist_pos(rng);
+		std::cout << "Poz " << temp_pos << "\n";
+
+		temp_protocol = (protocol_map[sum])[temp_pos];
+
+		int temp_length = temp_protocol.size();
+
+		std::cout << "Dlugosc " << temp_length << "\n";
+
+		std::sort(temp_protocol.begin(), temp_protocol.begin() + temp_length);
+
+		// Losuj permutację
+		std::random_shuffle ( temp_protocol.begin(), temp_protocol.end() );
+
+		// Losuj 2 czasy
+		// 72000 kroków symulacji co 6 sekund
+		// zaczynamy w pierwszym kroku nr 1
+		// między dawkami 30min = 300 kroków
+		// czyli:
+		// 1.1 weź długość dawek. dawkki mogą zacząć się najpóźniej 
+		//     w 72000 - długość * 300 kroku = 72000 - koniec
+		// 1.2 losuj pierwszy termin dawki z 1..72000-koniec
+		// 1.3 dodaj do niego 300 i losuj kolejny termin dawki z 
+		//     (poprzedni + 300, 14400-koniec + 300)
+		// 1.4 itd, aż nie dojdziesz do końca długości wektorów
+
+		std::vector<int> random_time_1;
+		std::vector<int> random_time_2;
+
+		int time_length = 0;
+
+		while (time_length < temp_length) {
+			std::uniform_int_distribution<int> dist_time(1, 72000);
+			int temp_time = dist_time(rng);
+			bool good_time = true;
+			for (int t = 0; t < random_time_1.size(); ++t) {
+				if ((std::abs(random_time_1[t] - temp_time)) < 300) {
+					good_time = false;
+				}
+			}
+			if (good_time) {
+				random_time_1.push_back(temp_time);
+				time_length++;
+			}
+		}
+
+		time_length = 0;
+
+		while (time_length < temp_length) {
+			std::uniform_int_distribution<int> dist_time(1, 72000);
+			int temp_time = dist_time(rng);
+			bool good_time = true;
+			for (int t = 0; t < random_time_2.size(); ++t) {

```

Comment:
Jeśli chcemy n dawek, to czy nie lepiej byłoby wylosować z liczb z zakresu od 0 do MAX_CZAS_DAWKI/300 i potem te wylosowane pomnożyć przez 300? Wtedy zagwarantowana byłaby ta odległość bez sprawdzania z każdą poprzednią dawką...

Reply:
Obgadane na mesku - tl;dr nie, bo wtedy bedziemy miec rowno co 300, bez flexibility interwalow.